File: src/skill_router/infra/scanner.py
"""Skill scanner — discovers and parses SKILL.md from configurable roots.

Default roots: ``~/.claude/skills``, ``~/.claude/plugins`` (recursive), and
``./.claude/skills`` for the current project. Recursively walks each root and
parses every ``SKILL.md`` it finds, deduplicating by absolute path.

Scope mapping:
  - "plugin"  — path contains ``/.claude/plugins/``
  - "user"    — path is under ``~/.claude/skills/``
  - "project" — path contains ``/.claude/skills/`` but is NOT under home
"""

from __future__ import annotations

import logging
import re
from pathlib import Path

import yaml

from skill_router.domain.types import ParsedSkill

logger = logging.getLogger(__name__)


def default_roots() -> list[Path]:
    """Return the list of default discovery roots that actually exist.

    Scope mapping for discovered paths:
      - ~/.claude/plugins/         -> "plugin"
      - ~/.claude/skills/          -> "user"
      - <project>/.claude/skills/  -> "project"  (auto-discovered from Claude registry)
      - SKILL_ROUTER_EXTRA_ROOTS   -> "project"  (colon-separated env var, isolated from Claude Code)
    """
    import os  # noqa: PLC0415
    user_skills = Path.home() / ".claude" / "skills"
    user_plugins = Path.home() / ".claude" / "plugins"
    candidates = [user_skills, user_plugins]
    candidates.extend(_discover_project_roots())

    extra_env = os.environ.get("SKILL_ROUTER_EXTRA_ROOTS", "")
    for raw in extra_env.split(":"):
        raw = raw.strip()
        if raw:
            candidates.append(Path(raw).expanduser())

    return [p for p in candidates if p.exists()]


def _decode_project_path(encoded: str) -> Path | None:
    """Decode a Claude project dir name to a filesystem path using greedy matching.

    e.g. '-home-catow-GIT-Hackathon' -> Path('/home/catow/GIT/Hackathon')
    Handles hyphens in directory names by checking filesystem existence at each step.
    """
    if not encoded.startswith('-'):
        return None
    parts = encoded[1:].split('-')
    current = Path('/')
    i = 0
    while i < len(parts):
        matched = False
        # Try longest match first (greedy) so 'decepticon-docs' beats 'decepticon' + 'docs'
        for j in range(len(parts), i, -1):
            candidate_name = '-'.join(parts[i:j])
            candidate = current / candidate_name
            if candidate.exists():
                current = candidate
                i = j
                matched = True
                break
        if not matched:
            # Remaining parts don't resolve to existing dirs — append as-is
            current = current / '-'.join(parts[i:])
            break
    return current


def _discover_project_roots() -> list[Path]:
    """Auto-discover project .claude/skills/ dirs from Claude Code's project registry."""
    registry = Path.home() / ".claude" / "projects"
    if not registry.exists():
        return []

    user_skills = Path.home() / ".claude" / "skills"
    found: list[Path] = []
    seen: set[Path] = set()

    for entry in sorted(registry.iterdir()):
        if not entry.is_dir():
            continue
        project_path = _decode_project_path(entry.name)
        if project_path is None:
            continue
        skills_dir = project_path / ".claude" / "skills"
        resolved = skills_dir.resolve() if skills_dir.exists() else skills_dir
        if skills_dir.exists() and resolved != user_skills.resolve() and resolved not in seen:
            seen.add(resolved)
            found.append(skills_dir)
            logger.debug("Discovered project skills: %s", skills_dir)

    logger.info("Auto-discovered %d project skill roots", len(found))
    return found


def _extra_project_roots() -> set[Path]:
    """Return resolved paths from SKILL_ROUTER_EXTRA_ROOTS env var."""
    import os  # noqa: PLC0415
    roots: set[Path] = set()
    for raw in os.environ.get("SKILL_ROUTER_EXTRA_ROOTS", "").split(":"):
        raw = raw.strip()
        if raw:
            p = Path(raw).expanduser().resolve()
            if p.exists():
                roots.add(p)
    return roots


def scope_for_path(path: Path, extra_project_roots: set[Path] | None = None) -> str:
    """Classify a SKILL.md path into 'project' | 'user' | 'plugin'."""
    # Use absolute (non-resolved) path so symlinked testbed dirs are classified by WHERE they live,
    # not where they ultimately resolve to.
    abs_str = str(path.absolute())
    resolved_str = str(path.resolve())
    home = str(Path.home().resolve())

    # Paths inside SKILL_ROUTER_EXTRA_ROOTS are always "project" — check unresolved path first
    if extra_project_roots:
        for root in extra_project_roots:
            root_s = str(root)
            if abs_str.startswith(root_s) or resolved_str.startswith(root_s):
                return "project"

    if "/.claude/plugins/" in resolved_str:
        return "plugin"
    if resolved_str.startswith(home + "/.claude/skills/") or resolved_str.startswith(home + "/.claude/skills"):
        return "user"
    if "/.claude/skills/" in resolved_str:
        return "project"
    return "user"


def _walk_skill_files(root: Path):
    """Yield all SKILL.md paths under root using os.walk (followlinks handles symlinked skill dirs)."""
    import os  # noqa: PLC0415
    for dirpath, _dirs, filenames in os.walk(root, followlinks=True):
        if "SKILL.md" in filenames:
            yield Path(dirpath) / "SKILL.md"


def scan_skills(roots: list[Path] | None = None) -> list[ParsedSkill]:
    """Recursively discover and parse all SKILL.md under the given roots.

    Parameters
    ----------
    roots:
        Directories to walk. ``None`` means use :func:`default_roots`.
    """
    if roots is None:
        roots = default_roots()

    extra_roots = _extra_project_roots()
    results: list[ParsedSkill] = []
    seen: set[Path] = set()

    for root in roots:
        if not root.exists():
            logger.warning("Root %s does not exist; skipping", root)
            continue
        for skill_path in _walk_skill_files(root):
            # Dedup by absolute path so symlinked testbed dirs aren't collapsed with their targets
            dedup_key = skill_path.absolute()
            if dedup_key in seen:
                continue
            seen.add(dedup_key)
            parsed = parse_skill_md(skill_path)
            if parsed is not None:
                parsed.scope = scope_for_path(skill_path, extra_roots)
                results.append(parsed)

    logger.info("Scanned %d SKILL.md files from %d roots", len(results), len(roots))
    return results


def scan_by_scope(roots: list[Path] | None = None) -> dict[str, list[ParsedSkill]]:
    """Scan skills and return them bucketed by scope.

    Returns a dict with keys "project", "user", "plugin", each holding a
    (possibly empty) list of ParsedSkill objects.
    """
    all_skills = scan_skills(roots)
    buckets: dict[str, list[ParsedSkill]] = {"project": [], "user": [], "plugin": []}
    for skill in all_skills:
        bucket = skill.scope if skill.scope in buckets else "user"
        buckets[bucket].append(skill)
    return buckets


def parse_skill_md(path: Path) -> ParsedSkill | None:
    """Parse a single SKILL.md.

    Returns ``None`` only if the file is completely unreadable. Malformed YAML,
    missing fields, etc. are handled gracefully via warnings.
    """
    try:
        raw = path.read_text(encoding="utf-8")
    except OSError as exc:
        logger.warning("Could not read %s: %s", path, exc)
        return None
    except UnicodeDecodeError as exc:
        logger.warning("Encoding error in %s: %s", path, exc)
        return None

    frontmatter: dict = {}
    body: str = raw
    warnings: list[str] = []

    if raw.startswith("---"):
        parts = raw.split("---", 2)
        if len(parts) >= 3:
            yaml_text = parts[1]
            body = parts[2].lstrip("\n")
            try:
                parsed_yaml = yaml.safe_load(yaml_text)
                if isinstance(parsed_yaml, dict):
                    frontmatter = parsed_yaml
                else:
                    warnings.append("YAML did not parse to a mapping")
                    frontmatter = _regex_frontmatter(yaml_text)
            except yaml.YAMLError as exc:
                warnings.append(f"YAML parse error: {exc}")
                frontmatter = _regex_frontmatter(yaml_text)
        else:
            warnings.append("Malformed front-matter: missing closing ---")

    # Resolve name (frontmatter -> parent directory fallback)
    raw_name = frontmatter.get("name")
    if raw_name:
        name = str(raw_name).strip()
    else:
        name = path.parent.name.lower().replace("_", "-")
        warnings.append("name missing; derived from parent directory")

    # Resolve description (frontmatter -> first body paragraph fallback)
    raw_description = frontmatter.get("description")
    if raw_description:
        description = " ".join(str(raw_description).split())
    else:
        first_para = _first_paragraph(body)
        if first_para:
            description = first_para
            warnings.append("description missing; derived from first body paragraph")
        else:
            description = ""

    return ParsedSkill(
        name=name,
        description=description,
        body=body,
        source_path=path,
        raw_frontmatter=frontmatter,
        warnings=warnings,
    )


def _regex_frontmatter(yaml_text: str) -> dict:
    """Best-effort ``key: value`` extraction when YAML parsing fails."""
    out: dict = {}
    for line in yaml_text.splitlines():
        m = re.match(r"^([A-Za-z][A-Za-z0-9_-]*)\s*:\s*(.+?)\s*$", line)
        if m:
            out[m.group(1)] = m.group(2)
    return out


def _first_paragraph(body: str) -> str:
    """Return the first non-empty, non-heading markdown paragraph (truncated)."""
    for chunk in re.split(r"\n\s*\n", body.strip()):
        chunk = chunk.strip()
        if chunk and not chunk.startswith("#"):
            return chunk[:500]
    return ""

File: src/infra/scanner.ts
import { existsSync, readdirSync, readFileSync, statSync } from "node:fs";
import { homedir } from "node:os";
import { resolve as resolvePath, sep, join, basename, dirname } from "node:path";

import { parse as parseYaml } from "yaml";

import type { ParsedSkill, SkillScope } from "../domain/types.js";

const HOME = homedir();
const HOME_RESOLVED = resolvePath(HOME);

function expanduser(p: string): string {
  if (p === "~") return HOME;
  if (p.startsWith("~/")) return join(HOME, p.slice(2));
  return p;
}

function safeStatExists(p: string): boolean {
  try {
    return existsSync(p);
  } catch {
    return false;
  }
}

function isDir(p: string): boolean {
  try {
    return statSync(p).isDirectory();
  } catch {
    return false;
  }
}

function decodeProjectPath(encoded: string): string | null {
  if (!encoded.startsWith("-")) return null;
  const parts = encoded.slice(1).split("-");
  let current = "/";
  let i = 0;
  while (i < parts.length) {
    let matched = false;
    // Greedy longest match first so 'decepticon-docs' beats 'decepticon' + 'docs'
    for (let j = parts.length; j > i; j--) {
      const candidateName = parts.slice(i, j).join("-");
      const candidate = join(current, candidateName);
      if (safeStatExists(candidate)) {
        current = candidate;
        i = j;
        matched = true;
        break;
      }
    }
    if (!matched) {
      // Remaining parts don't resolve; append as-is
      current = join(current, parts.slice(i).join("-"));
      break;
    }
  }
  return current;
}

function discoverProjectRoots(): string[] {
  const registry = join(HOME, ".claude", "projects");
  if (!safeStatExists(registry)) return [];

  const userSkills = resolvePath(join(HOME, ".claude", "skills"));
  const found: string[] = [];
  const seen = new Set<string>();

  let entries: string[];
  try {
    entries = readdirSync(registry).sort();
  } catch {
    return [];
  }

  for (const entry of entries) {
    const fullPath = join(registry, entry);
    if (!isDir(fullPath)) continue;
    const projectPath = decodeProjectPath(entry);
    if (projectPath === null) continue;
    const skillsDir = join(projectPath, ".claude", "skills");
    if (!safeStatExists(skillsDir)) continue;
    const resolved = resolvePath(skillsDir);
    if (resolved === userSkills) continue;
    if (seen.has(resolved)) continue;
    seen.add(resolved);
    found.push(skillsDir);
  }

  return found;
}

function extraProjectRoots(): Set<string> {
  const roots = new Set<string>();
  const raw = process.env.SKILLOGY_EXTRA_ROOTS ?? "";
  for (const part of raw.split(":")) {
    const trimmed = part.trim();
    if (!trimmed) continue;
    const expanded = resolvePath(expanduser(trimmed));
    if (safeStatExists(expanded)) roots.add(expanded);
  }
  return roots;
}

export function defaultRoots(): string[] {
  const userSkills = join(HOME, ".claude", "skills");
  const userPlugins = join(HOME, ".claude", "plugins");
  const candidates: string[] = [userSkills, userPlugins, ...discoverProjectRoots()];

  const raw = process.env.SKILLOGY_EXTRA_ROOTS ?? "";
  for (const part of raw.split(":")) {
    const trimmed = part.trim();
    if (trimmed) candidates.push(expanduser(trimmed));
  }

  return candidates.filter((p) => safeStatExists(p));
}

export function scopeForPath(path: string, extraRoots?: Set<string>): SkillScope {
  const abs = resolvePath(path);
  const resolved = resolvePath(path);
  const home = HOME_RESOLVED;

  if (extraRoots) {
    for (const root of extraRoots) {
      if (abs.startsWith(root) || resolved.startsWith(root)) return "project";
    }
  }

  if (resolved.includes(`${sep}.claude${sep}plugins${sep}`)) return "plugin";

  const userSkillsPrefix = home + `${sep}.claude${sep}skills`;
  if (resolved.startsWith(userSkillsPrefix + sep) || resolved === userSkillsPrefix) return "user";

  if (resolved.includes(`${sep}.claude${sep}skills${sep}`)) return "project";

  return "user";
}

function* walkSkillFiles(root: string): Generator<string> {
  let stack: string[] = [root];
  const seen = new Set<string>();
  while (stack.length) {
    const dir = stack.pop()!;
    let realDir: string;
    try {
      realDir = resolvePath(dir);
    } catch {
      continue;
    }
    if (seen.has(realDir)) continue;
    seen.add(realDir);

    let entries: string[];
    try {
      entries = readdirSync(dir);
    } catch {
      continue;
    }

    for (const name of entries) {
      const full = join(dir, name);
      let st;
      try {
        st = statSync(full);
      } catch {
        continue;
      }
      if (st.isDirectory()) {
        stack.push(full);
      } else if (name === "SKILL.md") {
        yield full;
      }
    }
  }
}

export function scanSkills(roots?: string[]): ParsedSkill[] {
  const useRoots = roots ?? defaultRoots();
  const extras = extraProjectRoots();
  const results: ParsedSkill[] = [];
  const seen = new Set<string>();

  for (const root of useRoots) {
    if (!safeStatExists(root)) continue;
    for (const skillPath of walkSkillFiles(root)) {
      const dedupKey = resolvePath(skillPath);
      if (seen.has(dedupKey)) continue;
      seen.add(dedupKey);
      const parsed = parseSkillMd(skillPath);
      if (parsed) {
        parsed.scope = scopeForPath(skillPath, extras);
        results.push(parsed);
      }
    }
  }

  return results;
}

export function scanByScope(roots?: string[]): Record<SkillScope, ParsedSkill[]> {
  const buckets: Record<SkillScope, ParsedSkill[]> = { project: [], user: [], plugin: [] };
  for (const skill of scanSkills(roots)) {
    const bucket = (["project", "user", "plugin"] as const).includes(skill.scope)
      ? skill.scope
      : "user";
    buckets[bucket].push(skill);
  }
  return buckets;
}

function regexFrontmatter(yamlText: string): Record<string, string> {
  const out: Record<string, string> = {};
  for (const line of yamlText.split(/\r?\n/)) {
    const m = /^([A-Za-z][A-Za-z0-9_-]*)\s*:\s*(.+?)\s*$/.exec(line);
    if (m) out[m[1]!] = m[2]!;
  }
  return out;
}

function firstParagraph(body: string): string {
  for (const chunk of body.trim().split(/\n\s*\n/)) {
    const trimmed = chunk.trim();
    if (trimmed && !trimmed.startsWith("#")) {
      return trimmed.slice(0, 500);
    }
  }
  return "";
}

export function parseSkillMd(path: string): ParsedSkill | null {
  let raw: string;
  try {
    raw = readFileSync(path, "utf-8");
  } catch {
    return null;
  }

  let frontmatter: Record<string, unknown> = {};
  let body = raw;
  const warnings: string[] = [];

  if (raw.startsWith("---")) {
    // Split into [pre, yaml, body] using the second '---'
    const parts = raw.split(/^---\s*$/m);
    // raw.split with ---\s*$/m on raw='---\nyaml\n---\nbody' produces ['', '\nyaml\n', '\nbody']
    if (parts.length >= 3) {
      const yamlText = parts[1] ?? "";
      body = (parts.slice(2).join("---")).replace(/^\n/, "");
      try {
        const parsedYaml = parseYaml(yamlText);
        if (parsedYaml && typeof parsedYaml === "object" && !Array.isArray(parsedYaml)) {
          frontmatter = parsedYaml as Record<string, unknown>;
        } else {
          warnings.push("YAML did not parse to a mapping");
          frontmatter = regexFrontmatter(yamlText);
        }
      } catch (exc) {
        warnings.push(`YAML parse error: ${(exc as Error).message}`);
        frontmatter = regexFrontmatter(yamlText);
      }
    } else {
      warnings.push("Malformed front-matter: missing closing ---");
    }
  }

  const rawName = frontmatter["name"];
  const name = typeof rawName === "string" && rawName.trim()
    ? rawName.trim()
    : (warnings.push("name missing; derived from parent directory"),
        basename(dirname(path)).toLowerCase().replace(/_/g, "-"));

  const rawDescription = frontmatter["description"];
  let description = "";
  if (typeof rawDescription === "string" && rawDescription.trim()) {
    description = rawDescription.split(/\s+/).filter(Boolean).join(" ");
  } else {
    const firstPara = firstParagraph(body);
    if (firstPara) {
      description = firstPara;
      warnings.push("description missing; derived from first body paragraph");
    }
  }

  return {
    name,
    description,
    body,
    sourcePath: path,
    rawFrontmatter: frontmatter,
    scope: "user",
    warnings,
  };
}
